### PR TITLE
Add addChild method to RequestLogBuilder

### LINKED ISSUE #869: RetryingHttpClient has to create a new context on each retry
`RetryingHttpClient` currently reuses the initial `ClientRequestContext` for retry requests. This has the following issues:

- `RequestLog` of the retries are not recorded.
- The `RequestLog` of the original request is recorded as 'failed' prematurely. We could record it as 'succeeded' if a retry succeeded.
- If there is a stateful `DecoratingClient`, which keeps its state in the context attribute map, between `RetryingHttpClient` and `HttpClientDelegate`, the `DecoratingClient` may not function correctly because the state of the retries may confuse it.

/cc @minwoox

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
`RetryingClient` could not record the logs properly.

Modifications:
- Add `addChild` to `RequestLogBuilder` to propagate the logs from child

Result:
- Close #869

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 1]
```diff
@@ -109,6 +113,67 @@ public DefaultRequestLog(RequestContext ctx) {
...
+            }
+            if (log.responseContent() != null) {
+                responseContent(log.responseContent(), log.rawResponseContent());
+            }
+            if (log.responseCause() != null) {
+                endResponse(log.responseCause());
+            } else {
+                endResponse();
+            }
+        }, COMPLETE);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: Doesn't need to use `COMPLETE` but just `RESPONSE_END`.

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 2]
```diff
@@ -612,13 +682,28 @@ private void notifyListeners(RequestLogListener[] listeners) {
...
+           .append(res)
+           .append('}');      // 1 char
+        if (children != null && children.size() > 1) {
+            buf.append("{totalAttempts=");
+            buf.append(children.size());
+            buf.append("}[");
+            for (int i = 0; i < children.size(); i++) {
+                buf.append(children.get(i));
+                if (i != children.size() - 1) {
+                    buf.append(',');
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: `", "`?

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLogBuilder.java [hunk 3]
```diff
@@ -36,6 +36,23 @@
...
+     * @see RequestLogBuilder#setRecentAsLastChild
+     */
+    void addChild(RequestLog child, boolean lastChild);
+
+    /**
+     * Sets the recent child as the last child log. The response-side logs of last child will be propagated.
+     * Note that the response-side logs are propagated when the {@link RequestLogAvailability} fulfills
+     * {@link RequestLogAvailability#COMPLETE}.
+     */
+    void setRecentAsLastChild();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLogBuilder.java]: `endResponseWithLastChild`?

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestLogBuilder.java [hunk 4]
```diff
@@ -36,6 +36,23 @@
...
+     *                  request-side logs will be propagated.
+     *
+     * @see RequestLogBuilder#setRecentAsLastChild
+     */
+    void addChild(RequestLog child, boolean lastChild);
+
+    /**
+     * Sets the recent child as the last child log. The response-side logs of last child will be propagated.
+     * Note that the response-side logs are propagated when the {@link RequestLogAvailability} fulfills
+     * {@link RequestLogAvailability#COMPLETE}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestLogBuilder.java]: `#RESPONSE_END`?

File: core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithLoggingTest.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                ctx.log().addListener(listener, RequestLogAvailability.REQUEST_END);
+                ctx.log().addListener(listener, RequestLogAvailability.COMPLETE);
+                return delegate().execute(ctx, req);
+            }
+        };
+    }
+
+    private static void assertRequestSideLog(RequestLog log) {
+        assertThat(log.requestHeaders()).isNotNull();
+        assertThat(log.requestHeaders().path()).isEqualToIgnoringCase("/hello");
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithLoggingTest.java]: `isEqualTo()`?

File: core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithLoggingTest.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        assertThat(log.toStringResponseOnly()).isEqualToIgnoringCase("{}"); // empty response
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithLoggingTest.java]: `isEqualTo()`?

File: core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java [hunk 7]
```diff
@@ -73,4 +86,60 @@ public void rpcFailure_endResponseDifferentCause() {
...
+        child.serializationFormat(SerializationFormat.NONE);
+        assertThat(log.serializationFormat()).isSameAs(SerializationFormat.NONE);
+
+        final HttpHeaders foo = HttpHeaders.of(AsciiString.of("foo"), "foo");
+        child.requestHeaders(foo);
+        assertThat(log.requestHeaders()).isSameAs(foo);
+
+        child.requestContent("baz", "bax");
+        assertThat(log.requestContent()).isSameAs("baz");
+        assertThat(log.rawRequestContent()).isSameAs("bax");
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java]: Could you extract the constants into variables so that it's less confusing?

File: core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java [hunk 8]
```diff
@@ -73,4 +86,60 @@ public void rpcFailure_endResponseDifferentCause() {
...
+                .isExactlyInstanceOf(RequestLogAvailabilityException.class);
+        assertThatThrownBy(() -> log.rawResponseContent())
+                .isExactlyInstanceOf(RequestLogAvailabilityException.class);
+
+        child.endResponse(new AnticipatedException("Oops!"));
+        // all response-side logs are propagated
+        assertThat(log.responseStartTimeMillis()).isEqualTo(child.responseStartTimeMillis());
+        assertThat(log.responseHeaders()).isSameAs(bar);
+        assertThat(log.responseContent()).isSameAs("baz1");
+        assertThat(log.rawResponseContent()).isSameAs("bax1");
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java]: Could you extract the constants into variables so that it's less confusing?

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 1 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: All of the codes are fixed. Thank you~!

File: core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java [hunk 9]
```diff
@@ -73,4 +86,58 @@ public void rpcFailure_endResponseDifferentCause() {
...
+        final String responseContent = "baz1";
+        final String rawResponseContent = "bax1";
+        child.responseContent(responseContent, rawResponseContent);
+        assertThat(log.responseContent()).isSameAs(responseContent);
+        assertThat(log.rawResponseContent()).isSameAs(rawResponseContent);
+
+        child.endResponse(new AnticipatedException("Oops!"));
+        final long childTotalDurationMillis = TimeUnit.NANOSECONDS.toMillis(child.totalDurationNanos());
+        assertThat(TimeUnit.NANOSECONDS.toMillis(log.totalDurationNanos()))
+                .isBetween(Math.max(0, childTotalDurationMillis - 300), childTotalDurationMillis); // 300 buffer
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java]: - 300 buffer -> Could you rephrase?
- Could you extract 300 into a variable? It's used twice.
- How about increasing to 1000 for less flakiness?

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 10]
```diff
@@ -109,6 +113,95 @@ public DefaultRequestLog(RequestContext ctx) {
...
+        if (children == null) {
+            // first child's all request-side logging events are propagated immediately to the parent
+            children = new ArrayList<>();
+            propagateRequestSideLog(child);
+        }
+        children.add(child);
+    }
+
+    private void propagateRequestSideLog(RequestLog child) {
+        child.addListener(log -> startRequest(log.channel(), log.sessionProtocol(), log.host()), REQUEST_START);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: Could we set the start time explicitly so that the start time of the parent request log is same with that of the first child? You could add a private variant of `startRequest()` which accepts the start time as a parameter.

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 11]
```diff
@@ -109,6 +113,95 @@ public DefaultRequestLog(RequestContext ctx) {
...
+        child.addListener(log -> serializationFormat(log.serializationFormat()), SCHEME);
+        child.addListener(log -> requestHeaders(log.requestHeaders()), REQUEST_HEADERS);
+        child.addListener(log -> requestContent(log.requestContent(), log.rawRequestContent()),
+                          REQUEST_CONTENT);
+        child.addListener(log -> {
+            if (log.requestCause() != null) {
+                endRequest(log.requestCause());
+            } else {
+                endRequest();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: Ditto - Could we set the end time explicitly so that the end time of the parent request log is same with that of the first child? You could add a private variant of `endRequest()` which accepts the start time as a parameter.

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 12]
```diff
@@ -109,6 +113,95 @@ public DefaultRequestLog(RequestContext ctx) {
...
+            endResponseFrom(lastChild);
+        }
+    }
+
+    private void startResponseFrom(RequestLog log) {
+        if (log instanceof DefaultRequestLog) {
+            startResponse0(true, ((DefaultRequestLog) log).responseStartTimeNanos,
+                           log.responseStartTimeMillis());
+        } else {
+            startResponse();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: Ditto

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 13]
```diff
@@ -109,6 +113,95 @@ public DefaultRequestLog(RequestContext ctx) {
...
+        if (lastChild.responseContent() != null) {
+            responseContent(lastChild.responseContent(), lastChild.rawResponseContent());
+        }
+    }
+
+    private void endResponseFrom(RequestLog log) {
+        if (log.responseCause() != null) {
+            endResponse(log.responseCause());
+        } else {
+            endResponse();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: Ditto

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 14]
```diff
@@ -82,11 +82,19 @@ public O execute(ClientRequestContext ctx, I req) throws Exception {
...
+        ctx.logBuilder().addChild(derivedContext.log());
         try (SafeCloseable ignore = RequestContext.push(derivedContext, false)) {
             return delegate().execute(derivedContext, req);
         }
     }
 
+    /**
+     * This should be called when retrying is finished.
+     */
+    protected static void onRetryingComplete(ClientRequestContext ctx) {
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Is there any reason making this method `static` unlike the other methods such as `setResponseTimeout`, `getNextDelay`, etc?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java [hunk 15]
```diff
@@ -120,16 +120,15 @@ protected HttpResponse doExecute(ClientRequestContext ctx, HttpRequest req) thro
...
         }
 
-        final HttpResponse response;
+        HttpResponse response;
         try {
             response = executeDelegate(ctx, rootReqDuplicator.duplicateStream());
         } catch (Exception e) {
-            closeOnException(deferredRes, rootReqDuplicator, e);
-            return;
+            response = HttpResponse.ofFailure(e);
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java]: Just curious. Does it aim to apply the retry strategy when catching an exception from the delegate?

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 16]
```diff
@@ -109,6 +113,121 @@ public DefaultRequestLog(RequestContext ctx) {
...
+            propagateRequestSideLog(child);
+        }
+        children.add(child);
+    }
+
+    private void propagateRequestSideLog(RequestLog child) {
+        child.addListener(log -> {
+            final long requestStartTimeNanos;
+            final long requestStartTimeMillis;
+            if (log instanceof DefaultRequestLog) {
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: nit: How about adding `requestStartTimeNanos()` and `requestEndTimeNanos()` to `RequestLog` interface? @trustin WDYT?

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 17]
```diff
@@ -109,6 +113,121 @@ public DefaultRequestLog(RequestContext ctx) {
...
+        checkState(!hasLastChild, "last child is already added");
+        checkState(!children.isEmpty(), "at least one child should be already added");
+        hasLastChild = true;
+        final RequestLog lastChild = children.get(children.size() - 1);
+        propagateResponseSideLog(lastChild);
+    }
+
+    private void propagateResponseSideLog(RequestLog lastChild) {
+        // update the available logs if the lastChild already has them
+        if (lastChild.isAvailable(RequestLogAvailability.RESPONSE_START)) {
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: RequestLogAvailability.RESPONSE_START -> RESPONSE_START? Because we have static imports already.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java (hunk 15 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java]: You are right. That is the reason.

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 17 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: Oops~ fixed

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java (hunk 16 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: That was my first thought. But I was not sure if it's ok to expose the value to the outside since it is relative value.
I'm not sticking to that though :-).

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 18]
```diff
@@ -612,13 +738,28 @@ private void notifyListeners(RequestLogListener[] listeners) {
...
-                  .append(req)
-                  .append(", res=") // 6 chars
-                  .append(res)
-                  .append('}')      // 1 char
-                  .toString();
+        buf.append("{req=")  // 5 chars
+           .append(req)
+           .append(", res=") // 6 chars
+           .append(res)
+           .append('}');      // 1 char
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java]: `//` is misaligned?

**Comment by trustin** [comment]
Thanks, @minwoox !
